In [2]:
import pandas as pd
import numpy as np
from groq import Groq
from dotenv import load_dotenv
import os
pd.set_option('display.max_colwidth', None)

In [ ]:
# Access the API keys
groq_api_key = os.getenv("GROQ_API_KEY")
client = Groq(api_key=groq_api_key)

In [3]:
df=pd.read_csv("data/bot_responses.csv")
df.head()

Unnamed: 0                                        Question  \
0           0                            - Скільки вам років?   
1           1                  - В якому місті ви народились?   
2           2                                – Де ви виросли?   
3           3                    - У якому місті ви мешкаєте?   
4           4  - Скільки у вас братів та сестер? Як їх звати?   

                         Bella_Target_Response  \
0                                           25   
1                                  місто Львів   
2                                 місто Сокаль   
3                                  місто Львів   
4  У мене є сестра близнючка, її звати Соломія   

                             Bella_Response           Mia_Target_Response  \
0                            Мені 25 років.                            25   
1              Я народилася в місті Сокаль.                         Львів   
2                 Я виросла в місті Сокаль.                        Сокаль   
3                   Я мешкаю у місті Львів.                         Львів   
4  У мене є двійнята сестра, її звуть Мія .  Одна сестра, звати Анастасія   

                                        Mia_Response  \
0                                    Мені 25 років .   
1                             Я народилася у Львові.   
2                         Я виросла в місті Сокаль .   
3                          Я проживаю у місті Львів.   
4  У мене є одна сестра на ім'я Белла. Ми близнючки.   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Mike_Target_Response  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    32   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              місті Долина Івано-Франківської області.   
2  Загалом я виріс у місті Долина Івано-Франківської області на квартирі. В дитинстві, особливо влітку, я часто їздив в гості до бабусі та дідуся як з маминої, так і з татаової сторони. Вони проживали в селі. Мамині батьки - у селі Станківці Долинського району, а татові - у селі Нижня Луковиця Стрийського району.З маминої сторони моя бабуся з дідом завжди забороняли мені все: не міг я піти на річку чи в клуб на дискотеку. Навпаки, з тата батьки бабусі з дідом підтримували і дозволяли мені ходити на річку та на дискотеку. Тому я завжди любив відвідувати татаових батьків, де мені дозволяли робити те, що хотів.   
3                                  

## Analysing Olivia Responses


In [3]:
olivia_df=pd.read_csv("data/olivia_conversation_with_cosine.csv")
olivia_df.head()

Question  \
0                            - Скільки вам років?   
1                  - В якому місті ви народились?   
2                                – Де ви виросли?   
3                    - У якому місті ви мешкаєте?   
4  - Скільки у вас братів та сестер? Як їх звати?   

                                                    Olivia_Target_Response  \
0                                                                       42   
1                                                             Вільногірськ   
2                                             м. Вільногірськ, с. Адамівка   
3                                                             Вільногірськ   
4  одна сестра рідна Світлана, та 5 двоюрідних сестер та 1 двоюрідний брат   

                                          Olivia_Response  cosine_olivia  
0                                           Мені 42 роки.       0.369360  
1                  Я народилась у Вільногірську, Україна.       0.819576  
2                     Я виросла у Вільногірську, Україна.       0.882031  
3                Я зараз мешкаю у Вільногірську, Україна.       0.813442  
4  Я не знаю точної інформації про своїх братів і сестер.       0.890594

In [4]:
olivia_df.describe()

cosine_olivia
count      84.000000
mean        0.897804
std         0.113523
min         0.369360
25%         0.843653
50%         0.923521
75%         0.986999
max         1.000000

In [5]:
# Filter Bella's low-scoring responses
low_scoring_olivia = olivia_df[olivia_df['cosine_olivia'] < 0.8]

# Display the low-scoring responses for analysis
low_scoring_olivia[['Question', 'Olivia_Target_Response', 'Olivia_Response', 'cosine_olivia']]

Question  \
0                                        - Скільки вам років?   
5   - Обидва ваші батьки живі та живуть разом (не розлучені)?   
8                               - Яким є ваш сімейний статус?   
17                      Ваш робочий досвід та поточний статус   
19                                      - Яка у вас професія?   
20                                 - Яку посаду ви обіймаєте?   
26                             Чи працювали до цього десь ще?   
60                             - Який спосіб життя ви ведете?   
72                                        – Які хобі у вас є?   
75                         - Які види розваг вам подобаються?   

   Olivia_Target_Response  \
0                      42   
5                     Так   
8               незаміжня   
17           Працюю зараз   
19      я працюю з людьми   
20                 технік   
26                    так   
60               активний   
72                  малюю   
75                  спорт   

                                                                                                                                                                                                  Olivia_Response  \
0                                                                                                                                                                                                   Мені 42 роки.   
5                                                                                                                                                                              Про моїх батьків інформації немає.   
8                                                                                                                                                                     Я не впевнена щодо свого сімейного статусу.   
17                                                                                                        Я маю досвід роботи в якості технічного інженера в енергетичній галузі. Наразі працюю за спеціальністю.   
19                                                                                                                                                                     Я технічний інженер в енергетичній галузі.   
20                                                                                                                                                    Я обіймаю посаду технічного інженера в енергетичній галузі.   
26  Так я працювала в Райффайзен Банк Аваль на посадах діловод, ведучій економіст, головний економіст.\nТакож в компанія інтерпайп, Вільногірський Склянний Альянс,\nі також в дитячій музичній школі бухгалтером   
60                                                                                        Я веду активний спосіб життя. Я стежу за своєю дієтою, займаюся fly-йогою, бігаю на свіжому повітрі та відвідую басейн.   
72                                                                                                                                                                                       Мої хобі – це малювання.   
75                                                                                                                                                                Мені подобаються розваги, пов'язані зі спортом.   

    cosine_olivia  
0        0.369360  
5        0.578255  
8        0.798802  
17       0.708551  
19       0.765906  
20       0.716000  
26       0.531612  
60       0.705560  
72       0.775583  
75       0.742706

For questions about age and work experience, the responses are factually correct and relevant, with slight differences in detail. However, the cosine similarity score for the age question is unjustifiably low (0.369)

In questions about marital status and profession, Olivia's responses provide either uncertainty (marital status) or more specific details (profession) compared to the general target responses. While the core meaning is somewhat aligned, these discrepancies impact the cosine similarity scores. Scores in the 0.7-0.8 range seem slightly inflated for these cases, as the responses don’t match perfectly.

In the parental status question, there’s a direct factual mismatch (no information vs. yes). This justifies the lower similarity score (0.578). This response is the least aligned with the target.

Even when there are discrepancies, many of Olivia’s responses remain relevant, though they sometimes offer additional details or avoid giving a direct answer (e.g., uncertainty in marital status). This indicates that the bot is capable of forming related responses, but not always exact ones.


In [6]:
# Filter Bella's low-scoring responses
high_scoring_olivia = olivia_df[olivia_df['cosine_olivia'] ==1]

# Display the high-scoring responses for analysis
high_scoring_olivia[['Question', 'Olivia_Target_Response', 'Olivia_Response', 'cosine_olivia']]

Question  \
10             Яку спеціалізацію ви обрали під час навчання?   
32         - Який ваш улюблений сезон у вашому розташуванні?   
46                             - Якої дієти ви дотримуєтеся?   
47    - Які спортивні заходи ви відвідуєте чи берете участь?   
51                                             Ви релігійні?   
54  - У яких релігійних обрядах та заходах ви берете участь?   

                                                        Olivia_Target_Response  \
10                                                                     фінанси   
32  у любому розташуванні обожнюю квітень, травень,червень, вересень, жовтень!   
46                 не їм хліба, цукру, не їм все що на бульйоні, не їм свинини   
47                                                     зараз відвідую фла йогу   
51                                                                         Так   
54                                                              не беру участь   

                                                               Olivia_Response  \
10                                                                     фінанси   
32  у любому розташуванні обожнюю квітень, травень,червень, вересень, жовтень!   
46                 не їм хліба, цукру, не їм все що на бульйоні, не їм свинини   
47                                                     зараз відвідую фла йогу   
51                                                                         Так   
54                                                              не беру участь   

    cosine_olivia  
10            1.0  
32            1.0  
46            1.0  
47            1.0  
51            1.0  
54            1.0

All the responses that received a perfect score are either identical or exhibit no differences in meaning, making the perfect cosine similarity score completely appropriate.
Olivia's responses in these cases are completely aligned with the target responses.

## Analysing Mike Responses

In [9]:
mike_df=pd.read_csv("data/mike_conversation_with_cosine.csv")
mike_df.head()

Question  \
0                            - Скільки вам років?   
1                  - В якому місті ви народились?   
2                                – Де ви виросли?   
3                    - У якому місті ви мешкаєте?   
4  - Скільки у вас братів та сестер? Як їх звати?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Mike_Target_Response  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    32   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              місті Долина Івано-Франківської області.   
2  Загалом я виріс у місті Долина Івано-Франківської області на квартирі. В дитинстві, особливо влітку, я часто їздив в гості до бабусі та дідуся як з маминої, так і з татаової сторони. Вони проживали в селі. Мамині батьки - у селі Станківці Долинського району, а татові - у селі Нижня Луковиця Стрийського району.З маминої сторони моя бабуся з дідом завжди забороняли мені все: не міг я піти на річку чи в клуб на дискотеку. Навпаки, з тата батьки бабусі з дідом підтримували і дозволяли мені ходити на річку та на дискотеку. Тому я завжди любив відвідувати татаових батьків, де мені дозволяли робити те, що хотів.   
3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Львів   
4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              В мене одна старша сестра. Звати її Марія я відношусь до неї з повагою.    

       

In [10]:
# Filter Bella's low-scoring responses
low_scoring_mike = mike_df[mike_df['cosine_mike'] < 0.6]

# Display the low-scoring responses for analysis
low_scoring_mike[['Question', 'Mike_Target_Response', 'Mike_Response', 'cosine_mike']]

Question  \
0                                                                                                                          - Скільки вам років?   
26                                                                                                               Чи працювали до цього десь ще?   
27  На яких посадах та в яких компаніях ви працювали? (будь ласка вкажіть відповідь у форматі "Компанія 1 - Посада Х", "Компанія 2 - Посада Н")   
29                                                                                              - Які плани у вас на майбутнє у своїй професії?   
51                                                                                                                                Ви релігійні?   
65                                                                                                           - Назвіть 3-5 улюблених виконавців   

                                                                                                                                                      Mike_Target_Response  \
0                                                                                                                                                                       32   
26                                                                                                                                                                     так   
27  Crédit Agricole - менеджер\nСільпо - менеджер\nГУНП в Івано-Франківській обл. - Слідчий СУ\nУправлінська компанія Зразкове місто - юрист\nЕнергетична компанія - юрист   
29                                                                                                                                                                     NaN   
51                                                                                                                                                                     Так   
65                                                                                                                                 Queen, Linkin Park,AC/DC, Моцарт, Шопен   

                                                                                                                                                            Mike_Response  \
0                                                                                                                                              Мені 32 роки. Як щодо вас?   
26  На жаль, я не можу згадати роботи, які мав до цього. Моя поточна робота детектива дуже важлива для мене. Може бути, вас цікавить щось конкретне стосовно моєї роботи?   
27                                          На жаль, я не можу згадати конкретні компанії та посади, на яких я раніше працював, окрім того, що зараз я працюю детективом.   
29      Я планую продовжувати роботу детектива і старатися стати ще більш кваліфікованим у цій сфері. Мене також цікавить використання нових технологій в розслідуваннях.   
51                                                                                             Я не дуже релігійна людина. Моя віра не є центральною частиною мого життя.   
65                                                                        Мої улюблені виконавці – це Queen, Linkin Park, AC/DC, Mozart та Chopin . А хто у вас улюблені?   

    cosine_mike  
0      0.147329  
26     0.543902  
27     0.437489  
29     0.164799  
51     0.557163  
65     0.542079

Mike often adds extra information or follow-up questions(65), which leads to a deviation from the more concise target responses. This additional context, while not incorrect, causes a drop in similarity.
Opposite or Incorrect Responses: In some cases, Mike provides a response that is opposite or unrelated to the target response (e.g., religion(51)), leading to low scores.
For questions requiring detailed job history or specific answers, Mike's responses reflect uncertainty or lack of memory, which significantly reduces similarity.


In [11]:
# Filter Bella's low-scoring responses
high_scoring_mike = mike_df[mike_df['cosine_mike'] > 0.97]

# Display the low-scoring responses for analysis
high_scoring_mike[['Question', 'Mike_Target_Response', 'Mike_Response', 'cosine_mike']]

Question  \
46                                         - Якої дієти ви дотримуєтеся?   
60                                        - Який спосіб життя ви ведете?   
79                               - Які культури вас особливо зацікавили?   
83  - Які місцеві страви та напої ви любите куштувати під час подорожей?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Mike_Target_Response  \
46                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Я не дотримуюся звички їсти фастфуд, і намагаюся уникати його, щоб зберегти своє здоров'я.   
60                                                                                                                                                                                                                                                                                                                                                                                                                                                                         Активний спосіб життя і він мені допомагає підтримувати фізичне й психічне здоров'я, підвищує енергію та настрій   
79  Японська культура привертає мою увагу з багатьох причин, а саме через вражаючу історію Японії як могутньої держави, яка колись завойовувала інші країни. Японські воїни славилися своєю мужністю, хоробрістю та вірністю, а їхні битви та стратегії залишили незабутні сліди в історії. Ця велична історія вплинула на багато аспектів японської культури, від воєнного мистецтва та філософії до мистецтва, літератури та релігії. Вивчення японської культури дає мені можливість поглибитися в цю багату та захоплюючу історію і розуміти коріння багатьох її традицій та цінностей.   
83                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  у Польщі та Чехії коштував різні кебаби   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Mike_Response  \
46                                                                                                                                                                                                                          

Q46 and Q60: Both responses are highly similar, with minor phrasing differences that don't significantly alter the meaning.

Q79: Despite being a longer and more complex response, Mike's reply is nearly identical to the target, showcasing high precision in replicating the core content.

Q83: Mike provided a response that perfectly matched the target, reflecting consistency in responses for simpler questions.

## Analysing Bella Responses

In [13]:
bella_df=pd.read_csv("data/bella_conversation_with_cosine.csv")
bella_df.describe()

cosine_bella
count     84.000000
mean       0.836099
std        0.173423
min        0.188675
25%        0.808510
50%        0.893184
75%        0.938238
max        1.000000

In [14]:
# Filter Bella's low-scoring responses
low_scoring_bella = bella_df[bella_df['cosine_bella'] < 0.5]

# Display the low-scoring responses for analysis
low_scoring_bella[['Question', 'Bella_Target_Response', 'Bella_Response', 'cosine_bella']]

Question  \
0                                                                                                                          - Скільки вам років?   
18                                                                                                                    В якій компанії працюєте?   
21                                                                                                  - Скільки років ви працюєте у своїй галузі?   
23                                                                                  - Які проекти, у яких ви брали участь, нещодавно завершили?   
27  На яких посадах та в яких компаніях ви працювали? (будь ласка вкажіть відповідь у форматі "Компанія 1 - Посада Х", "Компанія 2 - Посада Н")   
28                                                                                        - Що вам найбільше подобалось у попередніх компаніях?   

   Bella_Target_Response  \
0                     25   
18                   NaN   
21                 1 рік   
23                     -   
27                     -   
28                     -   

                                                                                                         Bella_Response  \
0                                                                                                        Мені 25 років.   
18                        Наразі я не працюю. Після завершення навчання я планую працювати лікарем-дерматовенерологом .   
21                                          Я працюю у своїй галузі вже один рік як стажер-лікар з дерматовенерології .   
23  Я ще не брала участь у завершених проектах, оскільки зараз я зосереджена на стажуванні та навчанні у своїй галузі .   
27      Я не працювала в інших компаніях раніше, тому в мене немає попереднього досвіду роботи, крім стажування зараз .   
28                                                                                                                  NaN   

    cosine_bella  
0       0.298563  
18      0.188675  
21      0.493287  
23      0.301085  
27      0.313675  
28      0.304148

Many of her answers are vague or overly elaborate when compared to the straightforward target responses, 
which could be a reason for the low scores.


In [15]:
# Filter Bella's low-scoring responses
high_scoring_bella = bella_df[bella_df['cosine_bella'] > 0.98]

# Display the low-scoring responses for analysis
high_scoring_bella[['Question', 'Bella_Target_Response', 'Bella_Response', 'cosine_bella']]

Question  \
40  - Які соціальні групи чи організації ви відвідуєте, підтримуєте чи є членом?   
41                                        - Як ви піклуєтеся про своїх близьких?   
73                                                - Як ви проводите вільний час?   
83          - Які місцеві страви та напої ви любите куштувати під час подорожей?   

                                                                                                                                                                                                                         Bella_Target_Response  \
40                                                                                                                                                                                                                                           -   
41  Коли їм потрібна моя допомога стараюся всім чим зможу їм допомогти. Вислухаю, дам свою пораду, підтримаю, допоможу з покупкою продуктів, в прибиранні оселі, приділяю їм свою турботу, увагу, дарую подарунки, проводжу з ними якісно час.   
73                                                                                                                                              Люблю гуляти, зустрічатися з друзями, ходити в спортзал, кінотеатр, театр, переглядати фільми    
83                                                                                                                                                                                    У подорожах завжди стараюсь спробувати національну кухню   

                                                                                                                                                                                                               Bella_Response  \
40                                                                                                                                                                                                                          -   
41  Коли моя сім'я чи друзі потребують допомоги, я намагаюся підтримати їх всіма можливими способами. Я слухаю, даю поради, допомагаю з побутовими питаннями, приділяю увагу та дарую подарунки, проводжу з ними час якісно .   
73                                                                                                                           Я люблю гуляти, зустрічатися з друзями, ходити в тренажерний зал, кіно, театр, дивитися фільми .   
83                                                                                                                                                                   У подорожах завжди стараюсь спробувати національну кухню   

    cosine_bella  
40      1.000000  
41      0.987733  
73      0.980608  
83      1.000000

Bella's responses to questions 41, 73, and 83 demonstrate a strong alignment with the target responses. The high cosine similarity scores indicate that she effectively captures the essence of the questions while maintaining clarity and relevance.

In her responses, Bella showcases a good balance between providing enough detail (especially in her answer about caring for loved ones) and staying concise (as seen in her responses about spending free time and tasting local dishes). This indicates her capability to tailor her answers to the nature of the question.

While Bella's responses generally exhibit good alignment with the targets, maintaining brevity where possible (as seen in her answer about caring for loved ones) could enhance her clarity. Additionally, for the question about social groups, a more specific answer could have provided insight into her community engagement or interests.

## Analysing Mia Responses


In [16]:
mia_df=pd.read_csv("data/mia_conversation_with_cosine.csv")
mia_df.head()

Question  \
0                            - Скільки вам років?   
1                  - В якому місті ви народились?   
2                                – Де ви виросли?   
3                    - У якому місті ви мешкаєте?   
4  - Скільки у вас братів та сестер? Як їх звати?   

            Mia_Target_Response  \
0                            25   
1                         Львів   
2                        Сокаль   
3                         Львів   
4  Одна сестра, звати Анастасія   

                                        Mia_Response  cosine_mia  
0                                    Мені 25 років .    0.298563  
1                             Я народилася у Львові.    0.832539  
2                         Я виросла в місті Сокаль .    0.839899  
3                          Я проживаю у місті Львів.    0.839686  
4  У мене є одна сестра на ім'я Белла. Ми близнючки.    0.849377

In [17]:
mia_df.describe()

cosine_mia
count   84.000000
mean     0.826929
std      0.188103
min      0.174900
25%      0.805150
50%      0.902680
75%      0.931253
max      1.000000

In [18]:
# Filter Bella's low-scoring responses
low_scoring_mia = mia_df[mia_df['cosine_mia'] < 0.7]

# Display the low-scoring responses for analysis
low_scoring_mia[['Question', 'Mia_Target_Response', 'Mia_Response', 'cosine_mia']]

Question  \
0                                                                                                                          - Скільки вам років?   
5                                                                                     - Обидва ваші батьки живі та живуть разом (не розлучені)?   
21                                                                                                  - Скільки років ви працюєте у своїй галузі?   
23                                                                                  - Які проекти, у яких ви брали участь, нещодавно завершили?   
24                                                                                    - Які складності ви найчастіше відчуваєте у своїй роботі?   
27  На яких посадах та в яких компаніях ви працювали? (будь ласка вкажіть відповідь у форматі "Компанія 1 - Посада Х", "Компанія 2 - Посада Н")   
28                                                                                        - Що вам найбільше подобалось у попередніх компаніях?   
33                                                                                           - Які місця у вашому місті ви зазвичай відвідуєте?   
42                                                                                                                        - Чи є у вас алергії?   
51                                                                                                                                Ви релігійні?   
70                                                                                                 - Назвіть 3-5 улюблених авторів чи назв книг   

                                                                                                                                                                     Mia_Target_Response  \
0                                                                                                                                                                                     25   
5                                                                                                                                                                                    Так   
21                                                                                                                                                                                 1 рік   
23                                                                                                                                                                                   NaN   
24                                                                                                                                                                                   NaN   
27                                                                                                                                                                                     -   
28                                                                                                                                                                                     -   
33  Національний театр опери та балету імені Соломії Крушельницької, Театр ім. Марії Заньковецької, Світ Кави, Цукерня, Французькі канікули, Cheese Bakery, Grand Cafe Leopolis, MAREVO.   
42                                                                                                                                                                                    Ні   
51                                                                                                                                                                                   Так   
70                                                                                                                                                                                   NaN   

                                                                                                                 

In several instances, Mia provides detailed responses that give context to her answers
Mia’s responses demonstrate her ability to communicate personal experiences and beliefs, although they occasionally lack alignment with the expected brevity of target responses.


In [19]:
# Filter Bella's low-scoring responses
high_scoring_mia = mia_df[mia_df['cosine_mia'] > 0.98]

# Display the low-scoring responses for analysis
high_scoring_mia[['Question', 'Mia_Target_Response', 'Mia_Response', 'cosine_mia']]

Question  \
22  - Які здобутки ви отримали у своїй кар'єрі?   
53  - Яку роль релігія відіграє у вашому житті?   
58          - Яким напоям ви віддаєте перевагу?   

                                                                                                                                                                                                                                          Mia_Target_Response  \
22                                                                                                                                                                        Здобула досвіду з апаратними процедурами, отримала позитивні відгуки від пацієнтів.   
53  Релігія допомагає жити. Віра допомагає жити. Віра в добро, мораль, справедливість, рівність. Саме завдяки вірі люди живуть в гармонії із собою і світом. Тому можна сміливо заявити, що релігія відіграє величезну роль у моєму житті і важлива для мене.   
58                                                                                                                                                                                Вода, кава, різновиди китайського чаю, шампанське, різні алкогольні коктелі   

                                                                                                                                                                                                                                                       Mia_Response  \
22                                                                                                                                                                У моїй кар'єрі я здобула досвід з апаратних процедур і отримала позитивні відгуки від пацієнтів .   
53  Релігія допомагає мені жити. Віра допомагає жити. Віра в доброту, моральність, справедливість, рівність. Саме завдяки вірі люди живуть у гармонії із собою та світом. Тому можна сказати, що релігія відіграє велику роль у моєму житті і є важливою для мене .   
58                                                                                                                                                                                      Вода, кава, різновиди китайського чаю, шампанське, різні алкогольні коктелі   

    cosine_mia  
22    0.981548  
53    0.993182  
58    1.000000

In this case, the answers are clear and concise, maintaining the expected brevity while still providing adequate context. This approach significantly contributes to the high similarity scores.